torchvision/models/maxvit.py

```diff
@@ -0,0 +1,595 @@
+        relative_bias = self.positional_bias[bias_index].view(self.max_seq_len, self.max_seq_len, -1)  # type: ignore
+        relative_bias = relative_bias.permute(2, 0, 1).contiguous()
+        return relative_bias.unsqueeze(0)
+
+    def forward(self, x: Tensor) -> Tensor:
+        # X, Y and stand for X-axis group dim, Y-axis group dim
+        B, G, P, D = x.shape
+        H, DH = self.n_heads, self.head_dim
+
+        qkv = self.to_qkv(x)
+        q, k, v = torch.chunk(qkv, 3, dim=-1)
+
+        q = q.reshape(B, G, P, H, DH).permute(0, 1, 3, 2, 4)
+        k = k.reshape(B, G, P, H, DH).permute(0, 1, 3, 2, 4)
+        v = v.reshape(B, G, P, H, DH).permute(0, 1, 3, 2, 4)
+
+        k = k * self.scale_factor
+        # X, Y and stand for X-axis group dim, Y-axis group dim
+        dot_prod = torch.einsum("B G H I D, B G H J D -> B G H I J", q, k)
+        pos_bias = self._get_relative_positional_bias()
+
+        dot_prod = F.softmax(dot_prod + pos_bias, dim=-1)
+
+        out = torch.einsum("B G H I J, B G H J D -> B G H I D", dot_prod, v)
+        out = out.permute(0, 1, 3, 2, 4).reshape(B, G, P, D)
+
+        out = self.merge(out)
+        return out
+
+
+class SwapAxes(nn.Module):
+    def __init__(self, a: int, b: int) -> None:
+        super().__init__()
+        self.a = a
+        self.b = b
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        res = torch.swapaxes(x, self.a, self.b)
+        return res
+
+
+class WindowPartition(nn.Module):
+    """
+    Function that takes in a tensor of shape [B, C, H, W] and partitions it
+    in to a tensor of shape [B, H/P, W/P, P*P, C]
+    """
+
+    def __init__(self, partition_size: int) -> None:
+        super().__init__()
+        self.partition_size = partition_size
+
+    def forward(self, x: Tensor) -> Tensor:
+        B, C, H, W = x.shape
+        P = self.partition_size
+        # chunk up H and W dimensions
+        x = x.reshape(B, C, H // P, P, W // P, P)
+        x = x.permute(0, 2, 4, 3, 5, 1)
+        # colapse P * P dimension
+        x = x.reshape(B, (H // P) * (W // P), P * P, C)
+        return x
+
+
+class WindowDepartition(nn.Module):
+    """
+    Function that takes in a tensor of shape [B, H/P, W/P, P*P, C]
+    and partitions it into a tensor of shape [B, C, H, W]
+    """
+
+    def __init__(self, partition_size: int, n_partitions: int) -> None:
+        super().__init__()
+        self.partition_size = partition_size
+        self.n_partitions = n_partitions
+
+    def forward(self, x: Tensor) -> Tensor:
+        B, G, PP, C = x.shape
+        P = self.partition_size
+        HP, WP = self.n_partitions, self.n_partitions
+        # split P * P dimension into 2 P tile dimensionsa
+        x = x.reshape(B, HP, WP, P, P, C)
+        # permute into B, C, HP, P, WP, P
+        x = x.permute(0, 5, 1, 3, 2, 4)
+        # reshape into B, C, H, W
+        x = x.reshape(B, C, HP * P, WP * P)
+        return x
+
+
+class MLP(nn.Module):
```

Comment:
There is an existing MLP block offered at TorchVision. See [here](https://github.com/pytorch/vision/blob/main/torchvision/ops/misc.py#L263).